### PR TITLE
add fmt and clippy tests as part of the test process

### PR DESCRIPTION
**Pull request by glommer**
So we don't have to have a separate CI target.
This is easier to reason, and it's a single test

### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Related issues

A list of issues either fixed, containing architectural discussions, otherwise relevant
for this Pull Request.

### Additional Notes

Anything else we should know when reviewing?

### Checklist

[] I have added unit tests to the code I am submitting
[] My unit tests cover both failure and success scenarios
[] If applicable, I have discussed my architecture

### CODE REVIEW

**Comment by duarten** [comment]
I feel this is harder to reason, as it replaces something simple, expected, and obvious (ci jobs) with something hidden and unexpected (tests calling into cargo) 😅

**Comment by glommer** [comment]
Then let's give others some time to comment before I debug more why this is failing.

But the reason I like this approach is that `cargo test` should be part of everyone's local development workflow.
Many a time I had run cargo test, sent the PR, just to see that the PR failed on fmt or clippy. Integrating those reduces the surprise factor.

**Comment by duarten** [comment]
> But the reason I like this approach is that `cargo test` should be part of everyone's local development workflow. Many a time I had run cargo test, sent the PR, just to see that the PR failed on fmt or clippy. Integrating those reduces the surprise factor.

True. Another approach for this would be to add a pre-commit hook.

**Comment by glommer** [comment]
Adding @HippoBaro @bryandmc @thirstycrow

**Comment by HippoBaro** [comment]
I believe this approach doesn't work because we rely on rust `nightly` for formatting. Since we use `stable` to run the tests, we end up with an incorrect result. 

> PS: Now that Rust 2021 has been released, maybe we should investigate whether we still need to use that weird setup. Is it possible the formatting options we use have been included in the stable channel?

**Comment by glommer** [comment]
Well, the approach itself works. The current code indeed doesn't, but I am one debugging session away from figuring out why =)

File: glommio/tests/linters.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#[cfg(test)]
+mod tests {
+    use std::process::Command;
+    #[test]
+    fn check_formating() {
+        let status = Command::new("cargo")
+            .args(["+nightly", "fmt", "--all", "--", "--check"])
+            .status()
+            .unwrap();
+        assert!(status.success());
```
Comment (HippoBaro) [inline glommio/tests/linters.rs:12]: Can you add a failed assertion message here? A common mistake is to format using the stable channel so a little hint there would help people understand what's going on!

**Review by HippoBaro — CHANGES_REQUESTED** [review]

**Comment by HippoBaro** [comment]
Overall I like the idea + it doesn't prevent people used to the fmt+clippy workflow to keep doing what they are currently used to

**Comment by HippoBaro** [comment]
Is there a reason why the `doc` test isn't part of this change? It would be nice for `cargo test` to fail if some APIs are not properly documented.

File: glommio/tests/linters.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                "-D",
+                "warnings",
+            ])
+            .status()
+            .unwrap();
+        assert!(status.success());
+    }
+
+    #[test]
+    fn check_sort() {
```
Comment (HippoBaro) [inline glommio/tests/linters.rs:39]: Can we rename this to something like `check_dependencies_sorted`?

Comment (glommer) [reply glommio/tests/linters.rs:39]: done

**Comment by glommer** [comment]
> Is there a reason why the `doc` test isn't part of this change? It would be nice for `cargo test` to fail if some APIs are not properly documented.

Not sure if that buys us much. I just removed a documentation from a public type, and got this:

```
$ cargo doc --all-features --no-deps
warning: missing documentation for an enum
   --> glommio/src/io/dma_file.rs:29:1
    |
29  | pub enum CloseResult {
    | ^^^^^^^^^^^^^^^^^^^^
    |
note: the lint level is defined here
   --> glommio/src/lib.rs:288:9
    |
288 | #![warn(missing_docs, missing_debug_implementations, rust_2018_idioms)]
    |         ^^^^^^^^^^^^

warning: `glommio` (lib doc) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
[glauber@trueserverstrongandfree glommio]$ echo $?
0
```

More importantly: 
1. both clippy and build complained about that as well.
2. `cargo doc` has side-effects (generates a whole doc directory). Not sure if this is advisable during test.

**Comment by HippoBaro** [comment]
Ah, I see, in that case, this LGTM! If that's okay with you, I'll merge when the test all passes (except for the one you removed).

**Comment by bryandmc** [comment]
maybe I should have mentioned this sooner but the other option is a git hook, right? Like a "push" hook that runs everything.. though people can always ignore them.. But it takes extra flags to their command.

**Comment by HippoBaro** [comment]
I would be fine with also adding a hook that runs `cargo test` as part of the PR; @glommer wdyt?

**Comment by glommer** [comment]
not sure what that buys us

**Comment by HippoBaro** [comment]
> maybe I should have mentioned this sooner but the other option is a git hook, right? Like a "push" hook that runs everything.. though people can always ignore them.. But it takes extra flags to their command.

IMO these two things are orthogonal, this PR can be merged and is useful on its own

**Comment by bryandmc** [comment]
> > maybe I should have mentioned this sooner but the other option is a git hook, right? Like a "push" hook that runs everything.. though people can always ignore them.. But it takes extra flags to their command.
> 
> IMO these two things are orthogonal, this PR can be merged and is useful on its own

Sorry misunderstood what this was for and only looked at the conversation not the code -- ignore me haha

**Pull request merged**